### DISCUSSION TITLE
Extends current enum to add new cases

### DISCUSSION THREAD

**Discussion by Jack-Works** [Jack-Works/proposal-enum#5] [General]
In https://github.com/microsoft/TypeScript/issues/17592 this feature is required

```ts
enum BasicEvents {
    Start = "Start",
    Finish = "Finish"
};

// extend enum using "extends" keyword
enum AdvEvents extends BasicEvents {
    Pause = "Pause",
    Resume = "Resume"
};
```

## Problems

This feature conflicts with Inline-able(#4) and hoistable declaration (#18).

It requires adding a initialize stage to "throw" when the extended `enum` is not a valid one.

It will trigger user code when the extended value is `getter` or `Proxy`.

Should we support this use case?

**Comment by rbuckton** [comment]
Enums in Python allow a restricted form of subclassing, where you can define methods on a base (or mix-in) class. You cannot, however, extend a base enum to add new enum members. 

In C#, you cannot extend an enum, but you can use extension methods to add related methods.

In TS, you cannot extend an enum via subclassing, but we do support declaration merging as long as the enums are in the same file (which I am *not* advocating for this proposal).

I don't believe subclassing enums should be considered for this proposal.

**Comment by masak** [comment]
Briefly: I'm in favor of enabling some kind of re-use of enum values between enums, but I think it sets up the wrong expectations to present it as "extending" the enum, and to use the `extends` keyword.

Here's why: `extends` for classes and interfaces means two things in practice:

1. You get all the relevant fields/properties/methods/values from the base type (what we want)
2. You get to be considered a subtype/derived type of the base type (very wrong in the case of enums)

The [README](https://github.com/Jack-Works/proposal-enum/blob/main/README.md#decided) already says that the enum is "non-extensible". This is the right default; [sum types](https://ncatlab.org/nlab/show/sum+type) like enums _should_ be non-extensible, because that's how you get value from matching/destructuring them. The second connotation of `extends` undermines that.
